### PR TITLE
CET 450 levels driven report and all scorecards report

### PR DESCRIPTION
**Pull request by cristinabuenahora**
## Change description

Three report types
Screen Shot 2022-02-14 at 3 34 00 PM

Levels driven report in Bird's eye with collapsible sections 
Screen Shot 2022-02-14 at 3 31 01 PM

## Type of change

- [ ] Bug fix (fixes an issue)
- [ ] New feature (adds functionality)
- [ ] Improvement (improves codebase without changing functionality)

## Related issues

> Fix [#1]()

### Development

- [ ] Lint rules pass locally
- [ ] Application changes have been tested thoroughly
- [ ] Automated tests covering modified code pass

### Security

- [ ] Security impact of change has been considered
- [ ] Code follows company security practices and guidelines

### Code review

- [ ] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached as necessary
- [ ] "Ready for review" label attached and reviewers assigned
- [ ] Changes have been reviewed by at least one other contributor
- [ ] Pull request linked to task tracker where applicable

### CODE REVIEW

File: src/components/ReportsPage/HeatmapPage/Tables/LevelsDrivenTable.tsx [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+import { HeatmapTableHeader } from './HeatmapTableHeader';
+import { LevelsInfoCell } from '../LevelsInfoCell';
+
+interface LevelsDrivenTableProps {
+  levels: string[];
+  groupBy: GroupByOption;
+  data: StringIndexable<ScorecardServiceScore[]>;
+}
+
+export const LevelsDrivenTable = ({
```
Comment (nikhilunni) [inline src/components/ReportsPage/HeatmapPage/Tables/LevelsDrivenTable.tsx:38]: this is too crazy not to share LOL
image

i have 3000 services locally so it gets to look super wild

Comment (nikhilunni) [reply src/components/ReportsPage/HeatmapPage/Tables/LevelsDrivenTable.tsx:38]: but very mesmerizing tbh

File: src/components/ReportsPage/HeatmapPage/Tables/LevelsDrivenTable.tsx [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  levels,
+  groupBy,
+  data,
+}: LevelsDrivenTableProps) => {
+  const notGroupedByServices = groupBy !== GroupByOption.SCORECARD;
```
Comment (nikhilunni) [inline src/components/ReportsPage/HeatmapPage/Tables/LevelsDrivenTable.tsx:43]: `GroupByOption.SCORECARD` should be `SERVICE`!

**Review by nikhilunni — APPROVED** [review]
just a couple nits, otherwise lgtm!

File: src/components/ReportsPage/HeatmapPage/Tables/LevelsDrivenTable.tsx (hunk 2 above)
Comment (nikhilunni) [reply src/components/ReportsPage/HeatmapPage/Tables/LevelsDrivenTable.tsx:43]: oh and similarly the display text should be by service:
image

File: src/components/ReportsPage/AllScorecardsPage/AllScorecardsPage.tsx [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+import React, { useCallback } from 'react';
+import { Content, ContentHeader } from '@backstage/core-components';
+import { Grid } from '@material-ui/core';
+import { AllScorecardsHeatmap } from './AllScorecardsHeatmap';
+import { useDropdown } from '../../../utils/hooks';
+import { GroupByOption } from '../../../api/types';
+import { GroupByDropdown } from '../Common/GroupByDropdown';
+import { CopyButton } from '../../Common/CopyButton';
+import { useLocation } from 'react-router';
+import { buildUrl } from '../../../utils/URLUtils';
```
Comment (kgrining) [inline src/components/ReportsPage/AllScorecardsPage/AllScorecardsPage.tsx:25]: let's keep the imports organized

File: src/components/ReportsPage/HeatmapPage/HeatmapPage.tsx [hunk 4]
```diff
@@ -63,16 +67,25 @@ export const HeatmapPage = () => {
...
         </Grid>
         <Grid item lg={12}>
           {selectedScorecardId === undefined ? (
-            <AllScorecardsHeatmap groupBy={groupBy!!} />
+            <EmptyState title="Select a scorecard" missing="data" />
           ) : (
             <SingleScorecardHeatmap
               scorecardId={selectedScorecardId}
               groupBy={groupBy!!}
+              headerType={headerType!!}
```
Comment (kgrining) [inline src/components/ReportsPage/HeatmapPage/HeatmapPage.tsx:88]: what does this double exclamation mark do?

File: src/components/ReportsPage/HeatmapPage/HeatmapPage.tsx [hunk 5]
```diff
@@ -63,16 +67,25 @@ export const HeatmapPage = () => {
...
+          </Grid>
+          <Grid item>
+            <HeaderTypeDropdown
+              headerType={headerType}
+              setHeaderType={setHeaderType}
+            />
+          </Grid>
         </Grid>
         <Grid item lg={12}>
           {selectedScorecardId === undefined ? (
```
Comment (kgrining) [inline src/components/ReportsPage/HeatmapPage/HeatmapPage.tsx:82]: let's use isUndefined from lodash if we have it

File: src/components/ReportsPage/HeatmapPage/HeatmapUtils.ts [hunk 6]
```diff
@@ -118,6 +129,29 @@ export const getAverageRuleScores = (
...
+  scores: ScorecardServiceScore[],
+  serviceCount: number,
+): number[] => {
+  return scores
+    .map(score =>
+      getSortedRulesByLevelsFromScores(sortedRulesByLevels, score).map(
+        rule => rule.score,
+      ),
+    )
+    .reduce((r, a) => a.map((b, i) => (r[i] || 0) + (b ? 1 : 0)), [])
```
Comment (kgrining) [inline src/components/ReportsPage/HeatmapPage/HeatmapUtils.ts:151]: ugh, this is hard to read, can we at least extract the mapper function and give it some meaningful name?

File: src/components/ReportsPage/HeatmapPage/Tables/HeatmapTableByLevels.tsx [hunk 7]
```diff
@@ -27,53 +26,34 @@ import {
...
-  } = useCortexApi(api => api.getScorecardLadders(scorecardId), [scorecardId]);
+  const rulesByLevels = getSortedRulesByLevels(rules, ladder?.levels);
 
-  if (loading) {
-    return <Progress />;
-  }
+  const headers = ['Level', 'Service Count', 'Average Score', ...rulesByLevels];
 
-  if (error || ladders === undefined) {
+  if (ladder === undefined) {
```
Comment (kgrining) [inline src/components/ReportsPage/HeatmapPage/Tables/HeatmapTableByLevels.tsx:46]: let's do isUndefined

File: src/components/ReportsPage/HeatmapPage/Tables/LevelsDrivenTable.tsx [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import React from 'react';
+import Table from '@material-ui/core/Table';
+import TableRow from '@material-ui/core/TableRow/TableRow';
+import { TableCell, Typography } from '@material-ui/core';
+import { GroupByOption, ScorecardServiceScore } from '../../../../api/types';
```
Comment (kgrining) [inline src/components/ReportsPage/HeatmapPage/Tables/LevelsDrivenTable.tsx:20]: let's organize imports

File: src/components/ReportsPage/HeatmapPage/Tables/SingleScorecardHeatmapTable.tsx [hunk 9]
```diff
@@ -14,50 +14,80 @@
...
+        : scores[0] && getSortedRuleNames(scores[0])) ?? [],
+    [levelsDriven, ladder, scores],
   );
+
   const data = useMemo(() => {
     return getScorecardServiceScoresByGroupByOption(scores, groupBy);
   }, [scores, groupBy]);
 
+  if (headerType === HeaderType.LEVELS) {
+    if (ladder === undefined) {
```
Comment (kgrining) [inline src/components/ReportsPage/HeatmapPage/Tables/SingleScorecardHeatmapTable.tsx:61]: isUndefined

File: src/components/ReportsPage/HeatmapPage/Tables/SingleScorecardHeatmapTable.tsx [hunk 10]
```diff
@@ -14,50 +14,80 @@
...
   scores,
+  ladder,
 }: SingleScorecardHeatmapTableProps) => {
-  const rules = useMemo(
-    () => (scores[0] && getSortedRuleNames(scores[0])) ?? [],
-    [scores],
+  const levelsDriven = headerType === HeaderType.LEVELS;
+  const headers = useMemo(
+    () =>
+      (ladder !== undefined && levelsDriven
```
Comment (kgrining) [inline src/components/ReportsPage/HeatmapPage/Tables/SingleScorecardHeatmapTable.tsx:50]: !isUndefined

File: src/utils/strings.ts [hunk 11]
```diff
@@ -26,3 +26,18 @@ const english_ordinal_rules = new Intl.PluralRules('en', { type: 'ordinal' });
 export function ordinal(value: number) {
   return suffixes[english_ordinal_rules.select(value)];
 }
+
+// Only handles cases where the plural form has a suffix appended. Does not handle cases like baby => babies, person => people
+export const maybePluralizeOnlyNoun = (
```
Comment (kgrining) [inline src/utils/strings.ts:31]: we should really start thinking about shared repo, these are great to be the first shared functions

**Review by kgrining — COMMENTED** [review]
mostly nits, otherwise lgtm

File: src/utils/strings.ts (hunk 11 above)
Comment (cristinabuenahora) [reply src/utils/strings.ts:31]: yepp... totally agree, maybe a Cortex npm utils package

File: src/components/ReportsPage/HeatmapPage/HeatmapPage.tsx (hunk 4 above)
Comment (cristinabuenahora) [reply src/components/ReportsPage/HeatmapPage/HeatmapPage.tsx:88]: basically just says it'll never be undefined (which it wont)

File: src/components/ReportsPage/HeatmapPage/Tables/LevelsDrivenTable.tsx (hunk 2 above)
Comment (cristinabuenahora) [reply src/components/ReportsPage/HeatmapPage/Tables/LevelsDrivenTable.tsx]: updated!

**Pull request merged**